alfresco-search/src/main/java/org/alfresco/solr/SolrInformationServer.java

```diff
@@ -4114,256 +2915,960 @@ public void reindexNodeByQuery(String query) throws IOException, AuthenticationE
+                    else
+                    {
+                        break;
+                    }
+                }
 
+                batchStartId = batchEndId + 1;
+                batchEndId = Math.min(batchStartId + BATCH_FACET_TXS, maxId);
             }
+
+            reporter.reportUniqueIdsInIndex(idsInIndex.cardinality());
         }
+    }
 
-        public LongHashSet getTxnSet() {
-            return txnSet;
+    private void setDuplicates(IndexHealthReport report, SolrQueryRequest request, String docType,
+                               SetDuplicatesCommand cmd)
+    {
+        // A mincount of 2 checks for duplicates in solr
+        NamedList<Integer> dbIdCounts = getFacets(request, FIELD_DOC_TYPE + ":" + docType, FIELD_DBID, 2);
+        for (Map.Entry<String, Integer> dbId : dbIdCounts)
+        {
+            long duplicatedDbId = Long.parseLong(dbId.getKey());
+            cmd.execute(report, duplicatedDbId);
         }
     }
 
-    class LRU extends LinkedHashMap
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private NamedList<Integer> getFacets(SolrQueryRequest request, String query, String field, int minCount)
     {
-        private int maxSize;
+        ModifiableSolrParams params =
+                new ModifiableSolrParams(request.getParams())
+                        .set(CommonParams.Q, query)
+                        .set(CommonParams.ROWS, 0)
+                        .set(FacetParams.FACET, true)
+                        .set(FacetParams.FACET_FIELD, field)
+                        .set(FacetParams.FACET_MINCOUNT, minCount);
 
-        public LRU(int maxSize) {
-            super((int)(maxSize*1.35));
-            this.maxSize = maxSize;
+        SolrQueryResponse response = cloud.getResponse(nativeRequestHandler, request, params);
+        NamedList facetCounts = (NamedList) response.getValues().get("facet_counts");
+        NamedList facetFields = (NamedList) facetCounts.get("facet_fields");
+        return (NamedList) facetFields.get(field);
+    }
+
+    private int getDocListSize(String query)
+    {
+        try (SolrQueryRequest request = this.newSolrQueryRequest())
+        {
+            ModifiableSolrParams params =
+                    new ModifiableSolrParams(request.getParams())
+                            .set(CommonParams.Q, query)
+                            .set(CommonParams.ROWS, 0);
+            ResultContext resultContext = cloud.getResultContext(nativeRequestHandler, request, params);
+            return resultContext.getDocList().matches();
         }
+    }
 
-        protected boolean removeEldestEntry(Map.Entry eldest) {
-            return size() > this.maxSize;
+    private String getFieldValueString(SolrDocument doc, String fieldName)
+    {
+        IndexableField field = (IndexableField)doc.getFieldValue(fieldName);
+        String value = null;
+        if (field != null)
+        {
+            value = field.stringValue();
         }
+        return value;
```

Comment:
We could use a ternary operator or a helper method that provides defaulting here?
```
return (field != null ? field.stringValue() : null);
```